src/lib/Libifl/pbsD_selectj.c

```diff
@@ -80,29 +178,71 @@ char **
 __pbs_selectjob(int c, struct attropl *attrib, char *extend)
 {
 	char **ret = NULL;
+	int i;
+	struct reply_list *rlist = NULL;
+	struct reply_list *cur;
+	svr_conn_t *svr_connections = get_conn_servers(c);
+	int num_cfg_svrs = get_num_servers();
+	int *failed_conn;
+	int rc = 0;
+
+	if (!svr_connections)
+		return NULL;
+
+	failed_conn = calloc(num_cfg_svrs, sizeof(int));
 
-	/* initialize the thread context data, if not already initialized */
 	if (pbs_client_thread_init_thread_context() != 0)
 		return NULL;
 
-	/* first verify the attributes, if verification is enabled */
-	if (pbs_verify_attributes(c, PBS_BATCH_SelectJobs, MGR_OBJ_JOB,
-		MGR_CMD_NONE, attrib))
+	if (pbs_verify_attributes(random_srv_conn(svr_connections), PBS_BATCH_SelectJobs, MGR_OBJ_JOB, MGR_CMD_NONE, attrib) != 0)
 		return NULL;
 
-	/* lock pthread mutex here for this connection */
-	/* blocking call, waits for mutex release */
 	if (pbs_client_thread_lock_connection(c) != 0)
 		return NULL;
 
-	if (PBSD_select_put(c, PBS_BATCH_SelectJobs, attrib, NULL, extend) == 0)
-		ret = PBSD_select_get(c);
+	for (i = 0; i <num_cfg_svrs; i++) {
 
-	/* unlock the thread lock and update the thread context data */
-	if (pbs_client_thread_unlock_connection(c) != 0) {
-		/* Even though ret is a char **, PBSD_select_get() allocated all its memory in one malloc() */
-		free(ret);
-		return NULL;
+		if (svr_connections[i].state != SVR_CONN_STATE_UP) {
+			pbs_errno = PBSE_NOSERVER;
+			continue;
+		}
+
+		if (pbs_client_thread_lock_connection(c) != 0)
+			goto done;
+
+		if ((rc = PBSD_select_put(svr_connections[i].sd, PBS_BATCH_SelectJobs, attrib, NULL, extend)) != 0) {
+			failed_conn[i] = 1;
+			continue;
+		}
+	}
+
+	for (i = 0; i < num_cfg_svrs; i++) {
+
+		if (svr_connections[i].state != SVR_CONN_STATE_UP)
+			continue;
+
+		if (failed_conn[i])
+			continue;
+
+		PBSD_select_get(svr_connections[i].sd, &rlist);
+
+	}
+
+	if (pbs_client_thread_unlock_connection(c) != 0)
+		goto done;
+
+	if (rc)
+		pbs_errno = rc;
+
+	free(failed_conn);
+
+done:
+	ret = reply_to_jobarray(rlist);
```

Comment:
can we rename this function from reply_to_jobarray() to reply_to_joblist() or something similar, so that it is not confued with array jobs?